### PR TITLE
Handle zero value modulus for OpenSSL 1.1

### LINKED ISSUE #78293: RSA.VerifyData terminates with "Out of memory." on Linux when passing an all-zero modulus
### Description

If an all-zero modulus is passed to RSA.VerifyData on Linux, the program will unexpectedly terminate instead of throwing an exception.

### Reproduction Steps

```cs
using System.Security.Cryptography;

var param = new RSAParameters{Modulus = new byte[256], Exponent = new byte[3]};
using (var rsa = RSA.Create(param))
{
    rsa.VerifyData(new byte[512], new byte[256], HashAlgorithmName.SHA256, RSASignaturePadding.Pss);
}
```

### Expected behavior

`VerifyData` throws a `CryptographicException` like it does on Windows.

### Actual behavior

The program terminates with `Out of memory. Command terminated by signal 6`

### Regression?

Issue occurs on .NET 6 and 7. Unsure about earlier versions.

### Known Workarounds

_No response_

### Configuration

I've tested running .NET 6.0 and 7.0 on both x64 Ubuntu 22.04 and Debian 11.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by vcsjones**
Fixes #78293

This PR introduces a check when importing an RSA public key that the modulus is not zero. For OpenSSL 1.1, the import succeeded and later fails at key usage time with a less-than-helpful error. This check brings consistency with other platforms, where a zero modulus fails at key import time.

### CODE REVIEW

**Comment by vcsjones** [comment]
@bartonjs this is what I ended up with as a fix for this for OpenSSL 1.x and would like your feedback on the overall approach.

`BN_is_zero` is a macro in OpenSSL 1.0. I _think_ I shimmed this correctly but would appreciate extra eyes over that.

**Comment by vcsjones** [comment]
> I _think_ I shimmed this correctly

Nope. Haha.

File: src/native/libs/System.Security.Cryptography.Native/opensslshim.h [hunk 1]
```diff
@@ -192,6 +192,7 @@ const EVP_CIPHER* EVP_chacha20_poly1305(void);
     REQUIRED_FUNCTION(BN_clear_free) \
     REQUIRED_FUNCTION(BN_dup) \
     REQUIRED_FUNCTION(BN_free) \
+    REQUIRED_FUNCTION(BN_is_zero) \
```
[the remark this answers is not in the document]
Comment (vcsjones) [reply src/native/libs/System.Security.Cryptography.Native/opensslshim.h]: @bartonjs do you recall an example of a 1.0.2 OpenSSL API that was implemented as a macro that later became a method that I could follow as a pattern? In the mean time, I guess I need to get building with OpenSSL 1.0.2 working again on my machine.

[the remark this answers is not in the document]
Comment (vcsjones) [reply src/native/libs/System.Security.Cryptography.Native/opensslshim.h]: Okay. `EVP_PKEY_CTX_set_rsa_keygen_bits` I think meets my criteria.

[the remark this answers is not in the document]
Comment (bartonjs) [reply src/native/libs/System.Security.Cryptography.Native/opensslshim.h]: I think `SSL_CTX_set_options` is in that bucket.

Function in 1.1.0: https://github.com/openssl/openssl/blob/OpenSSL_1_1_0-stable/include/openssl/ssl.h#L508

Macro in 1.0.2: https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/ssl/ssl.h#L785-L786

**Comment by vcsjones** [comment]
@bartonjs Seems like Windows 7 CSP doesn't fail at import, it fails at key usage. Given that

1. This only happens with a legacy implementation `RSACryptoServiceProvider`, not `RSACng` or `RSA.Create`
2. Windows 7 is very near EOL.

I opted to skip the test for Windows 7 instead of fixing RSACryptoServiceProvider. I'm not strongly opposed to fix it, but it would be adding Windows 7 specific code for an OS that is going to be EOL by the time .NET 8 is released.

**Review by stephentoub — APPROVED** [review]

**Review by bartonjs — APPROVED** [review]

**Comment by akoeplinger** [comment]
@vcsjones this seems to cause an error on the Linux x86 build:
```
 /__w/1/s/src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c:114:47: error: use of undeclared identifier 'EVP_R_INVALID_KEY'
                ERR_put_error(ERR_LIB_EVP, 0, EVP_R_INVALID_KEY, __FILE__, __LINE__);
```

I guess we have an older OpenSSL in that image? it uses `mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-cross-x86-linux`

**Comment by vcsjones** [comment]
@akoeplinger Looking.

**Comment by vcsjones** [comment]
> docker run -it mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-cross-x86-linux bash
root@9c730a75779d:/# openssl version
OpenSSL 1.1.1  11 Sep 2018

This is... confusing.

And it was definitely present in OpenSSL 1.1.1:

https://github.com/openssl/openssl/blob/OpenSSL_1_1_1a/include/openssl/evperr.h#L153

@bartonjs Any thoughts? It might be a while before I can really dig in to this. I suppose we can revert if this is blocking the build while I try to figure this out.

**Comment by bartonjs** [comment]
It's present in 1.1.0:  https://github.com/openssl/openssl/blob/OpenSSL_1_1_0-stable/include/openssl/evp.h#L1559 (# define EVP_R_INVALID_KEY                                163)

And 1.1.1: https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/openssl/evperr.h#L162 (still 163)

And 3.0: https://github.com/openssl/openssl/blob/openssl-3.0/include/openssl/evperr.h#L66

It's also present in 1.0.2: https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/evp/evp.h#L1587, though back then it was 171.

So... I'm fairly well confused.  Unless somehow we're building against 1.0.1 headers somewhere...

**Pull request merged**